File: protobuf-examples/src/main/java/tutorial/AddPerson.java
package tutorial;

import java.io.*;

public class AddPerson {
    // dice.Main function:  Reads the entire address book from a file,
    //   adds one person based on user input, then writes it back out to the same
    //   file.
    public static void main(String[] args) throws IOException {
        if (args.length != 1) {
            System.err.println("Usage:  AddPerson ADDRESS_BOOK_FILE");
            System.exit(-1);
        }

        AddressBook.Builder addressBook = AddressBook.newBuilder();

        // Read the existing address book.
        addressBook.mergeFrom(new FileInputStream(args[0]));

        // Add a person with an address
        addressBook.addPeople(
                promptForAddress(new BufferedReader(new InputStreamReader(System.in))));
        // Write the new address book back to disk.
        FileOutputStream output = new FileOutputStream(args[0]);
        addressBook.build().writeTo(output);
        output.close();
    }

    // This function fills in a Person message based on user input.
    private static Person promptForAddress(BufferedReader stdin) throws IOException {
        Person.Builder person = Person.newBuilder();
        System.out.print("Enter person ID: ");
        person.setId(Integer.parseInt(stdin.readLine()));

        System.out.print("Enter name: ");
        person.setName(stdin.readLine());

        System.out.print("Enter email address (blank for none): ");
        String email = stdin.readLine();
        if (email.length() > 0) {
            person.setEmail(email);
        }

        while (true) {
            System.out.print("Enter a phone number (or leave blank to finish): ");
            String number = stdin.readLine();
            if (number.length() == 0) {
                break;
            }
            Person.PhoneNumber.Builder phoneNumber =
                    Person.PhoneNumber.newBuilder().setNumber(number);
            System.out.print("Is this a mobile, home, or work phone? ");
            String type = stdin.readLine();
            if (type.equals("mobile")) {
                phoneNumber.setType(Person.PhoneType.MOBILE);
            } else if (type.equals("home")) {
                phoneNumber.setType(Person.PhoneType.HOME);
            } else if (type.equals("work")) {
                phoneNumber.setType(Person.PhoneType.WORK);
            } else {
                System.out.println("Unknown phone type.  Using default.");
            }

            person.addPhones(phoneNumber);
        }
        return person.build();
    }
}

File: protobuf-examples/src/main/kotlin/tutorial/add_person.kt
package tutorial

import tutorial.PersonKt.phoneNumber
import kotlin.io.path.Path
import kotlin.io.path.exists
import kotlin.io.path.inputStream
import kotlin.io.path.outputStream
import kotlin.system.exitProcess

// Reads the entire address book from a file, adds one person based
// on user input, then writes it back out to the same file.
fun main(arguments: Array<String>) {
    if (arguments.size != 1) {
        println("Usage: add_person ADDRESS_BOOK_FILE")
        exitProcess(-1)
    }
    val path = Path(arguments.single())
    val initialAddressBook = if (!path.exists()) {
        println("File not found. Creating new file.")
        addressBook {}
    } else {
        path.inputStream().use {
            AddressBook.newBuilder().mergeFrom(it).build()
        }
    }
    path.outputStream().use {
        initialAddressBook.copy { people += promptPerson() }.writeTo(it)
    }
}

fun promptPerson(): Person = person {
    print("Enter person ID: ")
    id = readlnOrNull()!!.toInt()

    print("Enter name: ")
    name = readln()

    print("Enter email address (blank for none): ")
    val email = readlnOrNull()
    if (email?.isNotEmpty()!!) {
        this.email = email
    }

    while (true) {
        print("Enter a phone number (or leave blank to finish): ")
        val number = readlnOrNull()
        if (number?.isEmpty()!!) break

        print("Is this a mobile, home, or work phone? ")
        val type = when (readlnOrNull()) {
            "mobile" -> Person.PhoneType.MOBILE
            "home" -> Person.PhoneType.HOME
            "work" -> Person.PhoneType.WORK
            else -> {
                println("Unknown phone type. Using home.")
                Person.PhoneType.HOME
            }
        }
        phones += phoneNumber {
            this.number = number
            this.type = type
        }
    }
}